pkg/patterns/addon/pkg/status/version.go

```diff
@@ -0,0 +1,81 @@
+package status
+
+import (
+	"context"
+	"fmt"
+	"github.com/blang/semver"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	addonsv1alpha1 "sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon/pkg/apis/v1alpha1"
+
+	//"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/manifest"
+)
+
+// NewVersionCheck provides an implementation of declarative.Reconciled that
+// checks the version of the operator if it is up to the version required by the manifest
+func NewVersionCheck(client client.Client, version string) *versionCheck {
+	return &versionCheck{client, version}
+}
+
+type versionCheck struct {
+	client  client.Client
+	version string
+}
+
+func (p *versionCheck) VersionCheck(
+	ctx context.Context,
+	src declarative.DeclarativeObject,
+	objs *manifest.Objects,
+) (bool, error) {
+	log := log.Log
+	zeroVersion := semver.Version{}
+	maxVersion := zeroVersion
+
+	// Look for annotation from any resource with the max version
+	for _, obj := range objs.Items {
+		annotations := obj.UnstructuredObject().GetAnnotations()
+		if versionNeededStr, ok := annotations["addons.k8s.io/operator-version"]; ok {
+			log.WithValues("version", versionNeededStr).Info("Got version, %v")
+
+			versionActual, err := semver.Make(versionNeededStr)
+			if err != nil {
+				log.WithValues("version", versionNeededStr).Info("Unable to convert string to version, skipping this object")
+				continue
+			}
+
+			if versionActual.GT(maxVersion) {
+				maxVersion = versionActual
+			}
+		}
+	}
+
+	// TODO(somtochi): Do we want to return an error when the version is invalid or just skip and use the operator?
+	operatorVersion, err := semver.Make(p.version)
+	if err != nil {
+		log.WithValues("version", p.version).Info("Unable to convert string to version, skipping check")
```

Comment:
You could move this to `NewVersionCheck`; I think it's very natural to return an error there if passed a bad version :-)

Related question is: should we refuse to apply manifests that specify invalid minimum versions (i.e. ignore vs error).  I would vote "error" I think, based on my chain of thoughts in this review!